README.md
# AlphaNumericCodeGenerator

Sencillo proyecto en `python 3` para generar secuencialmente todas las combinaciones de letras y números.
Inicialmente pensado para saber cuanto se tardaría en realizar esta acción y calcular el porcentage que representa solo la parte numérica

La idea no es _generar como tal todos los códigos_ sino ver cual es la mejor metodología para hacerlo y comprobar como influye en su rendimiento las diferentes arquitecturas de software o soluciones

### Pre-requisitos 📋

python 3.7 [Download](https://www.python.org/downloads/release/python-375/)


### Instalación 🔧

Clona o descarga el repositorio

Accede a la carpeta del proyecto y ejecuta 
```
python combinatoria.py
```

## Construido con 🛠️

* [Python](https://www.python.org/downloads/release/python-375/) - Operativo en cualquier arquitectura soportada


## Autores ✒️

* **Ezequiel Merino** - *Desarrollo y documentación* - [github](https://github.com/merinocabreraezequiel)

## Licencia 📄

Este proyecto está bajo la Licencia (GPL3) - mira el archivo [LICENSE](LICENSE) para detalles

## Agradecimientos 🎁

* Dr. Ricardo Zarzuela, por las consultas sobre la aproximación matemática de los valores con los que trabajamos



HISTORY.md
# ORGIEN
> Para un `Refactor` de un proyecto son surgió una problema. Migrabamos de un sistema por IDs a uno por Códigos.
> Como queriamos mantener las dos funcionalidades durante un tiempo, debiamos discernir si eran numero o códigos, pero, estos códigos eran alphanumeriocos, así que podía darse el caso de un Codigo que si fuese numerico y nos reventaria la solución.
> Pero, en un código de 12 digitos alphanumericos ¿Cuantos son solo numeros? ¿Vale la pena controlar esos casos?

# PRIMEROS CALCULOS
> Empezamos haciendo un sencillo calculo para saber cuantas combinaciones totales hay, resulta que son muchas **62^12**, esto sale de que, son 62 caracteres, todo el abecedario en mayusculas y minusculas más los 10 numeros, y es 12, porque decidimos que 12 digitos era un buen numero "solo 8 se terminarán muy pronto" dije, ***iluso***.
>> 3.2262668e+21, 3226266762397899821056, 3.2262667623979e+21, las calculadoras mas o menos se ponen de acuerdo
> Y de todos estos, cuantos son solo combinaciones numericas?
>> 1e+12, 1000000000000, 1000000000000, en esto todas estan de acuerdo 10^12 es manejable.
> Recuperando mastes de primaria, el tanto porciento debería ser `((10^12)*100)/(62^12)`
>> 3.0995577e-8, 0.00000003099557704449576, 0, vamos, que si no es 0, como dice una de las calculadoras, casi casi casi

# PRIMERAS CONCLUSIONES
> Resulta que es practicamente absurdo contemplar la opción, así que, si es numerico, damos por sentado que es un ID y si peta, ya hay un `try/catch` que se encarga de ello.

# GENERANDO CÓDIGOS
> Todo esto nos hizo preguntarnos cuanto tardariamos en sacar todas las combinaciones posibles desde `aaaaaaaaaaaa` hasta `999999999999`. La respuetsa fue sencilla, lanza un simple script que lo haga, secuencialmente y contando, en python, que se tadan dos minutos.
> Tras más de una semana generando códigos en una maquina dedicada, y no habiendo llegado aún al digito 10, nos dimos cuenta de que talvez habría que replantearlo...


combinatoria_v2.py
#Worst than the first
import string
import time

def dayshoursminutes():
	totalSeconds = round(time.time() - start_time)
	day = totalSeconds // (24 * 3600)
	totalSeconds = totalSeconds % (24 * 3600)
	hour = totalSeconds // 3600
	totalSeconds %= 3600
	minutes = totalSeconds // 60
	totalSeconds %= 60
	seconds = totalSeconds
	dhms = str(day) + ":" + str(hour) + ":" + str(minutes) + ":" + str(seconds)
	return (dhms)

start_time = time.time()
count=0
onlyDigit=0
lettersAndDigits = string.ascii_letters + string.digits
for a in lettersAndDigits:
	for b in lettersAndDigits:
		for c in lettersAndDigits:
			for d in lettersAndDigits:
				for e in lettersAndDigits:
					for f in lettersAndDigits:
						for g in lettersAndDigits:
							for h in lettersAndDigits:
								for i in lettersAndDigits:
									for j in lettersAndDigits:
										for k in lettersAndDigits:
											for l in lettersAndDigits:
												count += 1
												code = a+b+c+d+e+f+g+h+i+j+k+l
												if (code.isdigit()):
													onlyDigit += 1
												print (onlyDigit, "/", count, "(", dayshoursminutes(), ") -", code)


combinatoria_v4.py
import string
import time

def stepbystep(characters = [0,0,0,0,0,0,0,0,0,0,0,0]):
	lettersAndDigits = string.ascii_letters + string.digits
	return ''.join(lettersAndDigits[characters[letterPosition]] for letterPosition in range(12))

def dayshoursminutes():
	totalSeconds = round(time.time() - start_time)
	day = totalSeconds // (24 * 3600)
	totalSeconds = totalSeconds % (24 * 3600)
	hour = totalSeconds // 3600
	totalSeconds %= 3600
	minutes = totalSeconds // 60
	totalSeconds %= 60
	seconds = totalSeconds
	dhms = str(day) + ":" + str(hour) + ":" + str(minutes) + ":" + str(seconds)
	return (dhms)

start_time = time.time()
count=0
onlyDigit=0
a = 0
while a <= 61:
	b=0
	while b <= 61:
		c=0
		while c <= 61:
			d=0
			while d <= 61:
				e=0
				while e <= 61:
					f=0
					while f <= 61:
						g=0
						while g <= 61:
							h=0
							while h <= 61:
								i=0
								while i <= 61:
									j=0
									while j <= 61:
										k=0
										while k <= 61:
											l = 0
											while l <= 61:
												count += 1
												code = stepbystep([a,b,c,d,e,f,g,h,i,j,k,l])
												if (code.isdigit()):
													onlyDigit += 1
												print (onlyDigit, "/", count, "(", dayshoursminutes(), ") -", code)
												l += 1
											k += 1
										j += 1
									i += 1
								h += 1
							g += 1
						f += 1
					e += 1
				e += 1
			d += 1
		c += 1
	b += 1
a += 1

combinatoria.py
import string
import time

def stepbystep(characters = [0,0,0,0,0,0,0,0,0,0,0,0]):
	lettersAndDigits = string.ascii_letters + string.digits
	return ''.join(lettersAndDigits[characters[letterPosition]] for letterPosition in range(12))

def dayshoursminutes():
	totalSeconds = round(time.time() - start_time)
	day = totalSeconds // (24 * 3600)
	totalSeconds = totalSeconds % (24 * 3600)
	hour = totalSeconds // 3600
	totalSeconds %= 3600
	minutes = totalSeconds // 60
	totalSeconds %= 60
	seconds = totalSeconds
	dhms = str(day) + ":" + str(hour) + ":" + str(minutes) + ":" + str(seconds)
	return (dhms)

start_time = time.time()
count=0
onlyDigit=0
for a in range(62):
	for b in range(62):
		for c in range(62):
			for d in range(62):
				for e in range(62):
					for f in range(62):
						for g in range(62):
							for h in range(62):
								for i in range(62):
									for j in range(62):
										for k in range(62):
											for l in range(62):
												count += 1
												code = stepbystep([a,b,c,d,e,f,g,h,i,j,k,l])
												if (code.isdigit()):
													onlyDigit += 1
												if (count % 1000000 == 0):
													print (onlyDigit, "/", count, "(", dayshoursminutes(), ") -", code)
